app/test/integration_test.go

```diff
@@ -172,6 +173,80 @@ func (s *IntegrationTestSuite) TestMaxBlockSize() {
 	}
 }
 
+func (s *IntegrationTestSuite) TestSharesInclusionProof() {
+	require := s.Require()
+	val := s.network.Validators[0]
+
+	// generate 100 randomly sized txs (max size == 100kb)
+	txs, err := generateSignedWirePayForBlobTxs(val.ClientCtx, s.cfg.TxConfig, s.kr, -1, s.accounts...)
+	require.NoError(err)
+
+	hashes := make([]string, len(txs))
+
+	for i, tx := range txs {
+		res, err := val.ClientCtx.BroadcastTxSync(tx)
+		require.NoError(err)
+		require.Equal(abci.CodeTypeOK, res.Code)
+		hashes[i] = res.TxHash
+	}
+
+	// wait a few blocks to clear the txs
+	for i := 0; i < 50; i++ {
+		require.NoError(s.network.WaitForNextBlock())
+	}
+
+	for _, hash := range hashes {
+		txResp, err := queryTx(val.ClientCtx, hash, true)
+		require.NoError(err)
+		require.Equal(abci.CodeTypeOK, txResp.TxResult.Code)
+
+		// verify that the transaction inclusion proof is valid
+		require.True(txResp.Proof.VerifyProof())
+
+		// get the transaction shares
+		node, err := val.ClientCtx.GetNode()
+		require.NoError(err)
+		blockRes, err := node.Block(context.Background(), &txResp.Height)
+		require.NoError(err)
+		beginTxShare, endTxShare, err := prove.TxSharePosition(blockRes.Block.Txs, uint64(txResp.Index))
+		require.NoError(err)
+
+		// split the block to shares
+		rawShares, err := shares.Split(blockRes.Block.Data, true)
+		require.NoError(err)
+
+		// verify the transaction shares proof
+		txProof, err := prove.SharesInclusion(
+			rawShares,
+			blockRes.Block.SquareSize,
+			appconsts.TxNamespaceID,
+			beginTxShare,
+			endTxShare,
+		)
+		require.NoError(err)
+		require.NoError(txProof.Validate())
```

Comment:
[blocking]
can we actually use the rpc here?

Reply:
This is not an `rpc` call, this is executing the `Validate` function that is defined for the `TxProof` struct.

Reply:
Now, I have a question.  Concerning the proof struct:

```
message SharesProof {
  // data raw shares that are proven.
  repeated bytes data = 1;
  // shares_proof NMT proofs for the raw shares. Could be more than one
  // if the shares span more than one row.
  repeated NMTProof shares_proof = 2;
  // namespace_id namespaceID of the shares.
  bytes namespace_id = 3;
  // rows_proof binary merkle proofs, to the data root, of the rows
  // containing the raw shares.
  RowsProof rows_proof = 4;
```

Currently, it is self-sufficient. It contains everything it needs to verify.
What I mean by that, is that you don't need to specify a data root as parameter, to check the proof against. The data root is already included there.

Also, it contains the raw shares that need to be proved.

Thus, calling `Validate()`, will check if it is a valid proof.

Which is similar to how the `TxProof` is implemented:

```
// TxProof represents a Merkle proof of the presence of a transaction in the Merkle tree.
type TxProof struct {
	RowRoots []tmbytes.HexBytes  `json:"root_hash"`
	Data     [][]byte            `json:"data"`
	Proofs   []*tmproto.NMTProof `json:"proof"`
}
```

Contrarily to, for example, the tendermint implementation of a merkle proof:

```

// Verify that the Proof proves the root hash.
// Check sp.Index/sp.Total manually if needed
func (sp *Proof) Verify(rootHash []byte, leaf []byte) error {
	leafHash := leafHash(leaf)
	if sp.Total < 0 {
		return errors.New("proof total must be positive")
	}
	if sp.Index < 0 {
		return errors.New("proof index cannot be negative")
	}
	if !bytes.Equal(sp.LeafHash, leafHash) {
		return fmt.Errorf("invalid leaf hash: wanted %X got %X", leafHash, sp.LeafHash)
	}
	computedHash := sp.ComputeRootHash()
	if !bytes.Equal(computedHash, rootHash) {
		return fmt.Errorf("invalid root hash: wanted %X got %X", rootHash, computedHash)
	}
	return nil
}
```
Where, you need to specify the leaf and the root hash to verify it.

Thus, do the proofs that we generate need to be self-verifiable, or they should need extra information from outside?

I personally think that the former makes more sense, since it is a multiproof. But, I am not sure.

Reply:
> This is not an rpc call, this is executing the Validate function that is defined for the TxProof struct.

sorry, I meant that I think the important thing to test here is that the abci query is working via RPC. The rest seems like it could be a unit test.

Reply:
and I forget why we don't pass the root when validating. We at least need the user to verify that the hashes match or its not safe

Reply:
> sorry, I meant that I think the important thing to test here is that the abci query is working via RPC. The rest seems like it could be a unit test.

This is testing that the proofs are generated correctly for real blocks.

What do you mean by the ABCI query? if we're testing the RPC, then this test should be on Celestia-core side, because that's where we're receiving the RPC call. Otherwise, on Celestia-app side, we're just taking that request, parsing it and then returning the proof.

Reply:
[fa1cbcf](https://github.com/celestiaorg/celestia-app/pull/1072/commits/fa1cbcf0ea8363720293ebd3830d26db7823f5cf) @evan-forbes Please take a look when you have time